core/src/ipc.rs

```diff
@@ -131,3 +136,105 @@ impl<T> FromRawFd for IpcReceiver<T> {
         }
     }
 }
+
+/// Creates a pair of sockets that are meant for passing file descriptors to partitions.
+pub fn io_pair<T>() -> TypedResult<(IoSender<T>, IoReceiver<T>)>
+where
+    T: AsRawFd,
+{
+    let (tx, rx) = socketpair(
+        AddressFamily::Unix,
+        SockType::Datagram,
+        None,
+        SockFlag::empty(),
+    )
+    .typ(SystemError::Panic)?;
+    unsafe { Ok((IoSender::from_raw_fd(tx), IoReceiver::from_raw_fd(rx))) }
+}
+
+#[derive(Debug)]
+/// Internal data type for the IO resource sender
+pub struct IoSender<T> {
+    socket: UnixDatagram,
+    _p: PhantomData<T>,
+}
+
+impl<T> IoSender<T>
+where
+    T: AsRawFd,
+{
+    /// Sends a resource to the receiving socket.
+    pub fn try_send(&self, resource: T) -> TypedResult<()> {
+        let fds = [resource.as_raw_fd()];
+        let cmsg = [ControlMessage::ScmRights(&fds)];
+        // Have to send some bytes along with the ancillary data.
+        let iov = [IoSlice::new(&[0u8])];
+        let io_fd = self.socket.as_raw_fd();
+        sendmsg::<()>(io_fd, &iov, &cmsg, MsgFlags::empty(), None).typ(SystemError::Panic)?;
```

Comment:
Are you sure that the `?` is right here? It may be better to handle the errors manually, as certain sendmsg(2) errors are not critical and can indicate a retry, such as `EINTR` or `EAGAIN`.

Reply:
We should not see EAGAIN, since the socket is blocking and is not an internet socket. `EINTR` if the hypervisor received a signal. If we should retry the `sendmsg` depends on the signal, so this should be handled by the caller.